// @cloudbase/lcap-business-wxshop/index.js
/**
 * 错误码
 */
module.exports.errCode = require('./constant/errcode');

/**
 * 数据表名
 */
module.exports.db = require('./constant/db');

/**
 * 分类相关配置
 */
module.exports.category = require('./constant/category');

/**
 * 订单列表配置
 */
module.exports.order = require('./constant/order');

/**
 * 配置表相关配置
 */
module.exports.config = require('./constant/config');

/**
 * 公共函数
 */
const injectIoc = require('./util/injectIoc');
const genWxApi = require('./util/genWxApi');
const { makeId, makeSpuId, makeSkuId } = require('./util/makeId');
const { sleep } = require('./util/sleep');
const { genRetryFn } = require('./util/genRetryFn');
module.exports.util = {
  injectIoc,
  genWxApi,
  makeId,
  makeSpuId,
  makeSkuId,
  sleep,
  genRetryFn,
};

/**
 * 公共API
 */
const getWxToken = require('./api/getWxToken');
const describeWxAppInfo = require('./api/describeWxAppInfo');
const addInlineCategory = require('./api/addInlineCategory');
const getBindWxApp = require('./api/getBindWxApp');
const getDsColName = require('./api/getDsColName');
module.exports.api = {
  getWxToken,
  describeWxAppInfo,
  addInlineCategory,
  getBindWxApp,
  getDsColName,
};
