// @cloudbase/weda-node-server/dist/routes/datasource/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const call_datasource_1 = require("../../fns/call-datasource");
const config_1 = require("../../config");
const weda_node_sdk_1 = require("@cloudbase/weda-node-sdk");
/**
 * 调用数据源方法
 *  ctx.path 格式为 /<dataSourceName>/<methodName>
 */
const restDataSource = async function restDataSource(ctx) {
    if (ctx.method !== 'POST')
        ctx.throw(405, config_1.HTTP_CODE_MAP['405'].message);
    const requestBody = ctx.request.body || {};
    const [dataSourceName, methodName] = ctx.path.replace(/^\//, '').split('/');
    const envObj = (0, weda_node_sdk_1.getCloudbaseContext)(ctx.req._scfContext);
    const privateInfo = {
        isPrivate: envObj.IS_PRIVATE_ENV === 'true',
        tcbHost: envObj.SCF_ROUTE_TO_TCB_HOST,
    };
    const result = await (0, call_datasource_1.callDataSource)({
        isTestMode: requestBody.isTestMode,
        name: dataSourceName,
        methodName,
        params: requestBody.params,
        privateInfo,
    });
    // result 的值可能为 null,  koa 中将 ctx.body 设置为 null 即表示将body清空, 后续对其再设置将无效
    //  ref: https://github.com/koajs/koa/issues/998
    if (result !== null)
        ctx.body = result;
};
exports.default = restDataSource;
