// @cloudbase/weda-node-sdk/dist/http-method/auth/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getToken = void 0;
const remote_api_1 = require("../../remote-api");
const CACHED_TOKEN_RESPONSE = {};
// TOKEN 缓存时间 5 分钟
const MAX_CACHE_DURATION = 5 * 60 * 1000;
function getCachedToken(dsId) {
    const cache = CACHED_TOKEN_RESPONSE[dsId];
    if (!cache || (Date.now() - cache.updatedAt) > MAX_CACHE_DURATION)
        return;
    return cache.response;
}
/**
 * 根据数据源ID获取 token
 * https://tcloud-dev.oa.com/document/product/1505/70061?!preview&!document=1
 * @param dsId 数据源ID
 * @returns
 */
function getToken(dsId) {
    const cache = getCachedToken(dsId);
    if (cache)
        return Promise.resolve(cache);
    return (0, remote_api_1.fetchWeda)({
        action: 'DescribeConnectorToken',
        data: {
            dataSourceId: dsId,
        },
    }).then((response) => {
        const data = response.Data;
        CACHED_TOKEN_RESPONSE[dsId] = {
            updatedAt: Date.now(),
            response: data,
        };
        return data;
    });
}
exports.getToken = getToken;
