// @cloudbase/page-module/dist/cloudbase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initCloudbaseInstance = exports.initCloudbaseInstanceForWxClient = exports.cloudbase = exports.initCloudbaseInstanceForNodejsForCurrentEnv = exports.initCloudbaseInstanceForNodejs = void 0;
const utils_1 = require("./utils");
const kBaseSdk = 'wx-server-sdk';
const kRuntimeEnv = (0, utils_1.identifyRuntimeEnv)();
/* eslint-disable max-len */
function initCloudbaseInstanceForNodejs(env, options = {}) {
    (0, utils_1.checkNodeModuleDependencies)(kBaseSdk);
    /* eslint-disable @typescript-eslint/no-require-imports */
    const cloud = require(kBaseSdk);
    const cloudInstance = cloud.createNewInstance({
        ...options,
        env,
    });
    return cloudInstance;
}
exports.initCloudbaseInstanceForNodejs = initCloudbaseInstanceForNodejs;
function initCloudbaseInstanceForNodejsForCurrentEnv() {
    if (kRuntimeEnv === utils_1.RuntimeEnv.NODEJS) {
        (0, utils_1.checkNodeModuleDependencies)(kBaseSdk);
        /* eslint-disable @typescript-eslint/no-require-imports */
        const cloud = require(kBaseSdk);
        return initCloudbaseInstanceForNodejs(cloud.getWXContext().ENV);
    }
    return null;
}
exports.initCloudbaseInstanceForNodejsForCurrentEnv = initCloudbaseInstanceForNodejsForCurrentEnv;
exports.cloudbase = initCloudbaseInstanceForNodejsForCurrentEnv();
function initCloudbaseInstanceForWxClient(env, options = {}) {
    // 微信小程序环境
    const cloudInstance = new wx.cloud.Cloud({
        resourceEnv: env,
    });
    cloudInstance.init(options);
    return cloudInstance;
}
exports.initCloudbaseInstanceForWxClient = initCloudbaseInstanceForWxClient;
function initCloudbaseInstance(env, options = {}) {
    if (kRuntimeEnv === utils_1.RuntimeEnv.NODEJS) {
        return initCloudbaseInstanceForNodejs(env, options);
    }
    if (kRuntimeEnv === utils_1.RuntimeEnv.WX_CLIENT) {
        return initCloudbaseInstanceForWxClient(env, options);
    }
    /* eslint-disable max-len */
    throw new Error('[ERROR] Unknown runtime environment, please use in Node.js or WX_CLIENT environment');
}
exports.initCloudbaseInstance = initCloudbaseInstance;
