// @cloudbase/page-module/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloudbase = exports.isInvokeByServerSide = exports.isInvokeByCloudFunction = exports.callSelfDefinedMethod = exports.init = exports.pageModule = exports.PageModule = void 0;
const pagemodule_1 = require("./pagemodule");
var pagemodule_2 = require("./pagemodule");
Object.defineProperty(exports, "PageModule", { enumerable: true, get: function () { return pagemodule_2.PageModule; } });
/**
 * 全局单例的 PageModule 实例
 */
exports.pageModule = new pagemodule_1.PageModule();
/**
 * 初始化全局的单例 PageModule
 * @param moduleName - 模块名称
 * @param options    - 模块参数
 * @returns
 */
function init(moduleName, options = {}) {
    return exports.pageModule.init(moduleName, options);
}
exports.init = init;
var framework_1 = require("./framework");
Object.defineProperty(exports, "callSelfDefinedMethod", { enumerable: true, get: function () { return framework_1.callSelfDefinedMethod; } });
Object.defineProperty(exports, "isInvokeByCloudFunction", { enumerable: true, get: function () { return framework_1.isInvokeByCloudFunction; } });
Object.defineProperty(exports, "isInvokeByServerSide", { enumerable: true, get: function () { return framework_1.isInvokeByServerSide; } });
var cloudbase_1 = require("./cloudbase");
Object.defineProperty(exports, "cloudbase", { enumerable: true, get: function () { return cloudbase_1.cloudbase; } });
