// @cloudbase/page-module/src/utils.ts

/* eslint-disable @typescript-eslint/naming-convention */
export function isInWxEnv(): boolean {
  // window && window.__wxjs_environment
  return !!__wxConfig
}
/* eslint-enable @typescript-eslint/naming-convention */

export function isSupportCloudbase(): boolean {
  return !!wx.cloud
}

export function isNodeEnv(): boolean {
  return global?.process?.release?.name === 'node'
}

export enum RuntimeEnv {
  NODEJS = 'nodejs',
  WX_CLIENT = 'WX_CLIENT',
  UNKNOWN = 'unknown'
}

export function identifyRuntimeEnv(): RuntimeEnv {
  // 注意检查顺序：先检查是否在 Node.js 环境，再检查是否在小程序环境
  if (isNodeEnv()) {
    return RuntimeEnv.NODEJS
  }
  if (isInWxEnv() && isSupportCloudbase()) {
    return RuntimeEnv.WX_CLIENT
  }

  return RuntimeEnv.UNKNOWN
}

export function checkNodeModuleDependencies(pkgName: string) {
  try {
    /* eslint-disable @typescript-eslint/no-require-imports */
    require(pkgName)
  } catch (e) {
    if (e.code === 'MODULE_NOT_FOUND') {
      /* eslint-disable max-len */
      throw new Error(`[ERROR][${pkgName}] is required for Node.js environment, please install '${pkgName}' first.`)
    } else {
      throw e
    }
  }
}
